noether/src/edge_generator_server.cpp

```diff
@@ -0,0 +1,164 @@
+static const std::string EDGE_PATH_CONFIG_PARAM = "edge_path_config";
+
+class EdgeGeneratorServer
+{
+public:
+  EdgeGeneratorServer():
+    server_(nh_, GENERATE_EDGE_PATHS_ACTION, boost::bind(&EdgeGeneratorServer::executeAction, this, _1),false)
+  {
+
+  }
+
+  ~EdgeGeneratorServer()
+  {
+
+  }
+
+  bool run()
+  {
+    if(!loadConfig())
+    {
+      return false;
+    }
+    server_.start();
+    ROS_INFO("Edge Generator Server is ready ...");
+    return true;
+  }
+
+protected:
+
+  bool loadConfig()
+  {
+    using namespace XmlRpc;
+
+    ros::NodeHandle ph("~");
+    XmlRpcValue config;
+    if(!ph.getParam(EDGE_PATH_CONFIG_PARAM,config))
+    {
+      ROS_ERROR("Did not find the '%s' parameter", EDGE_PATH_CONFIG_PARAM.c_str());
+      return false;
+    }
+
+    try
+    {
+      edge_gen_config_.octree_res = static_cast<double>(config["octree_res"]);
+      edge_gen_config_.search_radius = static_cast<double>(config["search_radius"]);
+      edge_gen_config_.num_threads = static_cast<int>(config["num_threads"]);
+
+      edge_gen_config_.neighbor_tol = static_cast<double>(config["neighbor_tol"]);
+
+      edge_gen_config_.edge_cluster_min = static_cast<int>(config["edge_cluster_min"]);
+      edge_gen_config_.voxel_size = static_cast<double>(config["voxel_size"]);
+      edge_gen_config_.kdtree_epsilon = static_cast<double>(config["kdtree_epsilon"]);
+
+      edge_gen_config_.max_intersecting_voxels = static_cast<int>(config["max_intersecting_voxels"]);
+      edge_gen_config_.min_projection_dist = static_cast<double>(config["min_projection_dist"]);
+
+    }
+    catch(XmlRpcException& e)
+    {
+      ROS_ERROR("Fail to parse edge generation config file, error msg: %s", e.getMessage().c_str());
+      return false;
+    }
+    return true;
+  }
+
+  void executeAction(const noether_msgs::GenerateEdgePathsGoalConstPtr& goal)
+  {
+    using namespace noether_msgs;
+    noether_msgs::GenerateEdgePathsFeedback feedback;
+    noether_msgs::GenerateEdgePathsResult res;
+
+    if(!loadConfig())
+    {
+      ROS_WARN("Could not load configuration, using previous one");
+    }
+
+    ros::Timer feedback_timer = nh_.createTimer(ros::Duration(),[&](const ros::TimerEvent& evnt)
+    {
+      server_.publishFeedback(feedback);
+    });
+
+    for(std::size_t i = 0; i < goal->surface_meshes.size(); i++)
+    {
+      auto& mesh = goal->surface_meshes[i];
+      feedback.current_mesh_index = i;
+      boost::optional< std::vector<geometry_msgs::PoseArray> > edge_path_poses = edge_path_gen_.generate(mesh,edge_gen_config_);
+      if(!edge_path_poses)
+      {
+        res.validities.push_back(false);
+        res.edge_paths.push_back(ToolRasterPath());
+
+        if(!goal->proceed_on_failure)
+        {
+          res.success = false;
+          break;
+        }
+      }
+      else
+      {
+        ToolRasterPath edge_path;
+        edge_path.paths = edge_path_poses.get();
+        res.edge_paths.push_back(std::move(edge_path));
```

Comment:
I don't believe it is necessary to use std::move the compiler will leverage move semantics. If you compiled with c++17 and -Wredundant-move -Wpessimizing-move it would produce a warning.

Reply:
This currently builds with c++14

Reply:
It is not really the c++ version just that if you have the ability to build with c++17 you have the gcc version that provides the warning. The actual version of gcc is 9.